### DISCUSSION TITLE
To what extent does the reported memory bandwidth represent actual maximum system memory bandwidth?

### DISCUSSION THREAD

**Discussion by dagbdagb** [memtest86plus/memtest86plus#369] [General]
Mostly curious if the reported memory bandwidth is based on the SPD data, or actually measured performance.
I assume reporting *actual* memory clock frequency and latencies requires talking to the memory controller, and that this requires fairly CPU-specific code. 

Code may be written with various objectives in mind. Finding errors is certainly one. Writing to (or reading from) memory as fast as the hardware allows may or may not be a partial overlap with the objective of finding errors as fast as possible.

So:
Is there any code path in memtest86+ which is optimized *purely* for measuring the maximum performance (in terms of bandwidth) of the memory subsystem? I.e. disregarding looking for errors, and utilizing whatever fancy tricks the CPU supports. 

If my question is misguided, I apologize in advance.

**Comment by debrouxl** [comment]
* the reported bandwidth is based on a performance measurement, since SPD data is not always readable or valid :)
* correct, reporting actual memory clock frequency and latencies requires interfacing with the MC. The first bits of code to that effect were  added to the new memtest86+ code base over the past year, and are available in today's memtest86+ 7.00. Implementing support for more mobile, desktop and server processors' MCs is planned, and will come as Sam's time permits.
I doubt he'll go back in time beyond, say, x86-64-v2, i.e. Intel Nehalem (~15 years old) and AMD Bulldozer (~13 years old), as even reaching that point is already a whole lot of work, which requires a sufficient level and time of physical access to at least one working machine for every single processor / MC.
* memtest86+ does not currently contain code to measure maximum memory bandwidth performance, conditionally using newer instruction sets (SIMD, etc.) to hopefully achieve it. Other well-established user-space programs already do it, though the usual suspects were closed-source, last time I checked. The main branch of memtest86+ does not currently contain SIMD code, to begin with.

**Reply by dagbdagb** [reply]
Thank you for responding.

I just found and played a bit with [pmbw](https://github.com/bingmann/pmbw). Which clearly illustrates that actual memory bandwidth as exhibited by any single program very much depends on the access pattern, array size and number of cores/types of cores involved. And, in my case (12900K), that involving the E-cores pessimizes overall performance.

Still, could be interesting to have a 'single number' stating what is possible if the runtime environment, application and dataset can be adapted for optimum performance on the hardware at hand.  I am well aware of the concept of feature creep, so I have no expectations of memtest86+ filling this niche. 

But the "domain overlap" between testing for memory errors and testing for memory performance **does** appear to be significant. :-)

Thank you for memtest86+
